packages/builds/src/node/error-parser.ts

```diff
@@ -0,0 +1,139 @@
+    severity: string,
+    /** name of the file where the error occurs */
+    file: string,
+    /** location within the file. e.g. source code line number, column */
+    location: string,
+    /** text of the error or warning */
+    message: string,
+    /** reported error code, if any */
+    code: string
+}
+
+export interface IErrorMatcher {
+    // TODO: not all fields are yet used...
+    "name": string,
+    "label": string,
+    /** maps to the build command type? e.g. "typescript", "gcc/g++", ... */
+    "owner": string,
+    /** paths generated are absolute or relative? values: 'absolute' or 'relative' */
+    "fileLocation": string,
+    /** Regexp that matches errors/warnings */
+    "pattern": IErrorPattern
+}
+
+export interface IErrorPattern {
+    "patternName": string,
+    /** regexp that catches error */
+    "regexp": string,
+    /** in regexp match, group that represents the file */
+    "fileGroup": number,
+    /** in regexp match, group that represents the location */
+    "locationGroup": number,
+    /** in regexp match, group that represents the severity: "error", "warning" */
+    "severityGroup": number,
+    /** in regexp match, group that represents the error code, if present */
+    "codeGroup": number,
+    /** in regexp match, group that represents the error message */
+    "messageGroup": number
+}
+
+
+/**
+ * Class to parse build logs, to find instances of errors and warnings.
+ */
+decorate(injectable(), Events.EventEmitter);
+@injectable()
+export class ErrorParser extends Events.EventEmitter implements IErrorParser {
+
+    public parse(errorMatcher: IErrorMatcher, stream: NodeJS.ReadableStream): Promise<IParsedError[]> {
+        return new Promise<IParsedError[]>((resolve, reject) => {
+            try {
+                const errors: IParsedError[] = [];
+
+                // a log chunk has arrived
+                stream.on('data', (chunk: Buffer) => {
+                    this.doParse(chunk, errorMatcher).forEach(entry => {
+                        errors.push(entry);
+                    });
+                });
+
+                stream.on('error', (err: String) => {
+                    reject(err);
+                });
+
+                // finished parsing log, emit "done" event
+                stream.on('end', () => {
+                    this.emit('done', {});
+                    resolve(errors);
+                });
+
+            } catch (err) {
+                reject('Problem reading stream');
+            }
+        });
+    }
+
+    private doParse(buffer: Buffer, errorMatcher: IErrorMatcher): IParsedError[] {
+        const parsedEntries: IParsedError[] = [];
+
+        const regex: RegExp = new RegExp(errorMatcher.pattern.regexp, 'gm');
+        let match = undefined;
+
+        let fileAndPath = undefined;
+        while (match = regex.exec(buffer.toString())) {
+            // translate path from relative to absolute, if required
```

Comment:
Possibly we could use something like https://nodejs.org/dist/latest-v7.x/docs/api/readline.html ?